### PR TITLE
Upgrade/Replace to DTDL V3 enabled Parser

### PR DESCRIPTION
**Pull request by sai-prajeet**
This PR contains the below changes:
1. Added a new DTDL Parser Package version 1.0.52.
2. Removed the existing Parser Package (Microsoft.Azure.DigitalTwins.Parser)
3. Updated the System.Text.Json to latest stable version (7.0.3).
4. Created a new test project Generator.V3.Tests that contains DTDL models with version 3.
5. Renamed existing unit tests project to Generator.V2.Tests.
6. The Access level for the properties in the file CloudToDeviceMethodOptions.cs under Generator.V2.Tests.Generated Project is changed to internal from public on running the unit tests as the original file under custom folder (Generator.Template Project) is having the properties with access level internal.

### CODE REVIEW

**Review by tekgiant — CHANGES_REQUESTED** [review]

**Comment by amdawood** [comment]
@sai-prajeet, please add tests that generate v3 models

File: src/Generator/Helper.cs [hunk 1]
```diff
@@ -22,4 +22,30 @@ internal static bool IsObsolete(this DTNamedEntityInfo entity)
...
     };
+
+    internal static async IAsyncEnumerable<T> AsAsyncEnumerable<T>(this IEnumerable<T> input)
+    {
+        foreach (var value in input)
+        {
+            yield return value;
+        }
+
+        await Task.Yield();
```
[the remark this answers is not in the document]
Comment (sai-prajeet) [reply src/Generator/Helper.cs]: Thanks! Updated the code to use the package that has the extension method.

File: src/Generator/Helper.cs [hunk 2]
```diff
@@ -22,4 +22,20 @@ internal static bool IsObsolete(this DTNamedEntityInfo entity)
         AccessModifier.PrivateProtected => "private protected",
         _ => throw new Exception($"Invalid value for AccessModifier: {modifier}"),
     };
+
+    public static int IndexOf<T>(this IReadOnlyList<T> self, T elementToFind)
```
Comment (amdawood) [inline src/Generator/Helper.cs:26]: I'm failing to figure out what his method is needed for?

Comment (ms-mikeb) [reply src/Generator/Helper.cs:26]: Because the IReadOnlyInterface doesn't have an IndexOf method, although I agree, I don't see a usage for it.

Comment (sai-prajeet) [reply src/Generator/Helper.cs:26]: This is being used in the WriteContent method in EnumPropEntity class.

File: src/Generator/Helper.cs [hunk 3]
```diff
@@ -22,4 +22,20 @@ internal static bool IsObsolete(this DTNamedEntityInfo entity)
...
+    {
+        int i = 0;
```
Comment (amdawood) [inline src/Generator/Helper.cs:28]:
```suggestion
        int index = 0;
```

File: src/Generator/Helper.cs (hunk 2 above)
Comment (amdawood) [reply src/Generator/Helper.cs:26]: So why wasn't needed before? Did the new package change the behavior of `WriteContent()`?

Comment (sai-prajeet) [reply src/Generator/Helper.cs:26]: Previously it was just List<DTEnumValueInfo> but after Package change the type is IReadOnlyList<DTEnumValueInfo>

**Comment by sai-prajeet** [comment]
> 


> 

@microsoft-github-policy-service agree company="Microsoft"

File: src/Generator/Helper.cs (hunk 3 above)
Comment (sai-prajeet) [reply src/Generator/Helper.cs]: updated the code

**Comment by sai-prajeet** [comment]
> @sai-prajeet, please add tests that generate v3 models
updated the test models context version to 3 and all the test cases are passing

**Review by amdawood — APPROVED** [review]
Please ensure the description is updated to reflect all the changes

**Comment by sai-prajeet** [comment]
> Please ensure the description is updated to reflect all the changes

Thanks Amr. Updated the PR description

File: src/Generator/Generator.csproj [hunk 4]
```diff
@@ -43,7 +43,8 @@
 
   <ItemGroup>
     <Using Include="Azure.DigitalTwins.Core" />
-    <Using Include="Microsoft.Azure.DigitalTwins.Parser" />
+    <Using Include="DTDLParser.Models" />
```
Comment (apoorvg) [inline src/Generator/Generator.csproj:46]: The new parser does not use MS.Azure namepsace but is using DTDL namespace? Or this is completely a different one.

Comment (sai-prajeet) [reply src/Generator/Generator.csproj:46]: The existing package Microsoft.Azure.DigitalTwins.Parser has been deprecated and they have recommended to use the DTDLParser.Models package for V2 & V3 Models.

File: test/Generator.Tests/TestDtdlModels/Amenities.json
Comment (tekgiant) [inline test/Generator.Tests/TestDtdlModels/Amenities.json:1]: The suggestion from Amr was to ADD tests and models for V3. Distinguishing a separate set of tests and models is important for being able to test V3 specific features that will be implemented after this while also being able to confirm we haven't broken anything with V2.

**Review by tekgiant — CHANGES_REQUESTED** [review]
This PR should be renamed to _"Upgrade/Replace to DTDL V3 enabled Parser"_

File: test/Generator.Tests/TestDtdlModels/Amenities.json
Comment (sai-prajeet) [reply test/Generator.Tests/TestDtdlModels/Amenities.json:1]: Added a new test project that contains V3 Models.

File: test/Generator.V3.Tests/AssertHelper.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (Shuvajyoti) [inline test/Generator.V3.Tests/AssertHelper.cs:1]: Are these helper files same for both v2 and v3 unit test projects? If yes, we can have a single copy in a common project. Let me know your thoughts on this.

Comment (sai-prajeet) [reply test/Generator.V3.Tests/AssertHelper.cs:1]: As discussed, there are only 2 helper files and moving them to the Generator Project would not be correct as these are related to unit test cases. Hence for now keeping them here.

**Review by sagunase — APPROVED** [review]

**Review by apoorvg — APPROVED** [review]

File: test/Generator.V3.Tests/AssertHelper.cs
Comment (ms-mikeb) [inline test/Generator.V3.Tests/AssertHelper.cs:1]: We shouldn't be duplicating every class for the V3 & V2 tests. Either a single project should handle generating and testing for the 2 different versions (my suggestion) or you add a common test project that the V3 and V2 projects can have a dependency on for some of these common classes that you're duplicating for no reason.

**Review by ms-mikeb — CHANGES_REQUESTED** [review]

File: test/Generator.V3.Tests/AssertHelper.cs
Comment (sai-prajeet) [reply test/Generator.V3.Tests/AssertHelper.cs:1]: Yes, removed the files. Using the files from V2 Tests Project.

File: Generator.V3.Tests.Generated/Generator.V3.Tests.Generated.csproj
[the remark this answers is not in the document]
Comment (sai-prajeet) [reply Generator.V3.Tests.Generated/Generator.V3.Tests.Generated.csproj:1]: Thanks for the catch. Removed the duplicate file.

File: test/Generator.V3.Tests/Generator.V3.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
+    <IsPackable>false</IsPackable>
+
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
```
Comment (ms-mikeb) [inline test/Generator.V3.Tests/Generator.V3.Tests.csproj:13]: This whole ItemGroup block shouldn't be needed.

Comment (ms-mikeb) [reply test/Generator.V3.Tests/Generator.V3.Tests.csproj:13]:
```
<ItemGroup>
    <Content Include="TestDtdlModels\*.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </Content>
  </ItemGroup>
  ```
  
  You'll want this just like it's done in the V2 project

**Review by ms-mikeb — CHANGES_REQUESTED** [review]

File: test/Generator.V3.Tests/Generator.V3.Tests.csproj (hunk 6 above)
Comment (sai-prajeet) [reply test/Generator.V3.Tests/Generator.V3.Tests.csproj:13]: fixed.

**Review by ms-mikeb — APPROVED** [review]

**Comment by Shuvajyoti** [comment]
Please fix the build.

**Comment by sai-prajeet** [comment]
> Please fix the build.

Fixed the Build issue

**Review by ms-mikeb — APPROVED** [review]

**Pull request merged**